ODLA/platforms/odla_popart/odla_compute.cc

```diff
@@ -94,210 +58,236 @@ odla_status odla_SetComputationItem(odla_computation comp, odla_item_type type,
-  auto opts = SessionOptions();
-
-  // Create InferenceSession
-  auto proto = g_comp->builder->getModelProto();
-  auto session = popart::InferenceSession::createFromOnnxModel(
-      proto, data_flow, device, popart::InputShapeInfo(), *opts);
-  *context = new _odla_context(g_comp, std::move(session));
-
-  // Compile graph, create engine and load into the IPU
-  // use compileAndExport() to frozen engine to specified path
-  (*context)->session->prepareDevice();
-  // Init seed
-  (*context)->session->setRandomSeed(0);
-  // Copy weights from host to IPU
-  (*context)->session->weightsFromHost();
-
+  //std::cout << "---> odla_CreateContext()" << std::endl;
+  _odla_computation::instance(false)->init(); // Place the init here to avoid long execution problem
+  *context = new _odla_pipeline_context(_odla_computation::instance());
+  //std::cout << "<--- odla_CreateContext()" << std::endl;
   return ODLA_SUCCESS;
 }
 
 odla_status odla_DestroyContext(odla_context ctx) {
-  delete (ctx);
+  //std::cout << "---> odla_DestroyContext()" << std::endl;
+  if(nullptr != ctx && ctx->hold("odla_DestroyContext"))
+    delete (ctx);
+  else
+    ;//std::cerr << "Encounter a odla_DestroyContext with null ctx" << std::endl;
+  //std::cout << "<--- odla_DestroyContext()" << std::endl;
   return ODLA_SUCCESS;
 }
 
 odla_status odla_DestroyComputation(odla_computation comp) {
-  // g_comp.reset();
+  //std::cout << "Mark the computation done ..." << std::endl;
+  comp->mark_done();
+  // int cnt = 0;
+  // while(comp->thread_complete_){
+  //   //std::cout << "[" << cnt << "] Wating for the thread loop end..." << std::endl;
+  //   std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  //   if(cnt++ > 50)
+  //   {
+  //     std::cerr << "Thread did not exit in 5 seconds, stop waiting..." << std::endl;
+  //     break;
+  //   }
+  // }
+  if (comp->session != nullptr){
+    comp->session->getDevice().getDeviceInfo()->detach();
+    comp->session.reset();
+    assert(comp->session == nullptr);
+  }
   return ODLA_SUCCESS;
 }
 
 odla_status odla_ExecuteComputation(odla_computation comp, odla_context context,
                                     odla_compute_mode mode,
                                     odla_device device) {
-  // Config StepIO
-  std::map<popart::TensorId, popart::IArray&> inputs;
-  for (auto& input : comp->inputs) {
-    inputs.emplace(input.first, *input.second);
-  }
-  std::map<popart::TensorId, popart::IArray&> outputs;
-  for (auto& output : comp->outputs) {
-    outputs.emplace(output.first, *output.second);
-  }
-
-  popart::StepIO stepio(inputs, outputs);
-  // Run on ipu
-  context->session->run(stepio);
+  //std::cout << "---> odla_ExecuteComputation()" << std::endl;
+  if(!context->hold("odla_ExecuteComputation"))
+    return ODLA_FAILURE;
+  comp->executor()->compute(comp, context, mode, device);
+  //std::cout << "<--- odla_ExecuteComputation()" << std::endl;
   return ODLA_SUCCESS;
```

Comment:
pipeline mode needs to be compatible with non-pipeline mode，consider add a flag， such as  
```cpp
if (comp->opts.enable_pipeline) {
        // do pipeline;
} else {
        // normal;
}
```